### ISSUE TITLE
Unwanted dependency conflict resolution beyond version bounds

### ISSUE THREAD

**Issue by xtracoder** [gradle/gradle#1014]
I’m not sure if this a new feature request or already a bug of not correctly working functionality _new in 3.2.x_. I’ve already asked for help here https://discuss.gradle.org/t/unwanted-upgrade-of-dependency-version/20622, but there were no replies.  Anyway – everything is below:

### Background
We have large run-time modular Java-based solution – different modules are developed by different teams in following infrastructure (most probably rather common):
* There is shared (for all teams) build environment with shared repository for generated modules/JARs.
* Deployment environments are – (`local` – per each team) -> (`QA` | `Staging` - shared for all) -> Prod

The problems arises in following conditions:
* `Team-2` develops `module-team-2` which  uses `module-team-1` and after testing locally it generates ‘release’ build and deploys their modules into `QA` container alone with the other modules already deployed there, but it fails to deploy ...
* … because Team-1 over ‘last time’ already has generated some new artifacts on shared build environment/repository (and is testing them in `local servers`) - this way `module-team-2` __during build__ is linked to ‘future’ version of `module-team-1`, which is not yet deployed to `QA` or `Staging`

In our `build.gradle` we refer dependencies via ranges and it happens that during version resolution Gradle takes future versions  beyond specified range – here is how output of ‘gradle dependencies’ looks like

    org.app:app-ui:[4.17.0,4.17.1) -> 4.19.0.01948

Because of that, generated artifacts use gradle-resolved version as a lower bound for dependencies and therefore fail to start on `QA` | `Staging` until Team #1 deploy their future version there (matter of 1-2 weeks). Error is like `Could not find matching version=[4.19.0,5) ..., available versions 4.17.0.02002.`

### Expected Behavior
Version resolution should use available versions **within** specified range, i.e.

    org.app:app-ui:[4.17.0,4.17.1) -> 4.17.0.99999

### Current Behavior
Gradle takes version beyond specified version ranges

    org.app:app-ui:[4.17.0,4.17.1) -> 4.19.0.01948

Note: I tried using (new in 3.2 ?) ` resolutionStrategy.preferProjectModules() ` - nothing changed. Actually from its documentation (https://docs.gradle.org/current/dsl/org.gradle.api.artifacts.ResolutionStrategy.html#org.gradle.api.artifacts.ResolutionStrategy:preferProjectModules()) I did not get clear understanding of what it does. I have feeling that this directive should ‘force’ versions that are defined __explicitly__ in compile dependencies in module currently being built, but it seems it is not.

### Context
Described in ‘Background’

### Steps to Reproduce (for bugs)
Unfortunately I’m not sure  how to reproduce that problem outside of our build environment.

### Your Environment
Gradle + Java + Jenkins as build env + OSGI container (Karaf) as runtime env.

**Comment by xtracoder** [comment]
After thinking a bit more, I would say that correct version resolution will be following:

* for 'compile' time (i.e. during build) - __select lowest__ version across all requested dependencies - this way needed API will be available for compilation. Out of this versions plugins (like bnd.tools) generated run-time version range according to configured policy - typically it is next major version from the currently selected one, e.g.: version 1.1 -> range [1.1,2)
* for run-time (during 'exec', 'test') - __select highest__ version among the  matching versions, to provide latest and (hopefully) most correct run-time implementation. Runt-time version should not go beyond upper range specified in the module being executed.

I guess some new implementation needs to appear here
https://github.com/gradle/gradle/blob/master/subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/DefaultArtifactDependencyResolver.java#L116

**Comment by oehme** [comment]
Can you please provide a small reproducible example? If this was the only dependency then Gradle should pick one inside the range. My best guess is that some other dependency is transitively pulling in the newer version. Gradle uses the newest version on conflicts. You can tell Gradle to fail on version conflicts instead to find problems like these earlier.

**Comment by xtracoder** [comment]
Re: _**My best guess is that some other dependency is transitively pulling in the newer version**_

As far as I understand - this happens exactly because of this, we are getting something like following

* cur_target_module 
  -> depends-on: `module_a:[1.1,1.2) -> 1.1999`
 .   -> depends-on: `module_b:[1.2,1.3)`
* cur_target_module 
  -> depends-on: `module_b:[1.1,1.2)`

I assume you know that notation uses `[` for inclusive and `)` for exclusive version-range. So it happens that there is no valid intersections of requested version ranges for `module_b` across dependencies of multiple modules. This triggers conflict resolution and ... selecting 'the newest' version of `module_b` **for compile time** of `cur_target_module` is not good approach (because of above mentioned reasons).

I can create an artificial sample in nearest few days if this description is not sufficient.


Re: _**You can tell Gradle to fail on version conflicts instead to find problems like these earlier**_

'failOnConlicts' triggers too many faults among 3rd party libraries and they have to be resolved one-by-one to fail on the next one. After 'forcing' versions  for 10+ 3rd party libraries, I've starting searching for another way and it appears that gradle has very handy task for this case - 'dependencyInsight' - it provides rather good report for a single dependency.

**Comment by oehme** [comment]
Thanks for the clarification, this is exactly what I thought. There are different ways around this. One of them is failing on version conflicts. The other one would be writing a small plugin that locks in the lower bound of dependencies requested directly by the user. That way transitive ones can't overwrite them.

In the long run opening up a conflict resolution api could help with these kinds of problems.

**Comment by xtracoder** [comment]
Re: **The other one would be writing a small plugin that locks in the lower** 

Can provide more insight regarding how this plugin should technically work? From my quick look onto sources of 3.2.1 it seems that version resolution is rather hardcoded in Gradle's internals, i.e. when and where can I plug in for custom conflict resolution?

**Comment by oehme** [comment]
You can access what the user specified using `configuration.allDependencies` and then use that in `configuration.dependencyResolution.eachDependency {}`. This would allow you to get the strategy you want at least for first-level user-defined dependencies. It doesn't work if there are conflicts between two transitive ones.

The latter would require hooking into conflict resolution, which as you found correctly is not extensible yet.

**Comment by xtracoder** [comment]
By enumerating through `eachDependency` I guess the only thing I would be able to do is adding a `force` to some specific version, but from me that will require enumeration of versions available in configured repositories - and I'm not aware about Gradle API which provides that information.

**Comment by bigdaz** [comment]
Hey @xtracoder, in Gradle 4.3 we improved our handling of version ranges. Now, Gradle will make a best effort to find a version that fits into all requested version ranges. 
https://docs.gradle.org/current/release-notes.html#version-ranges-are-now-handled-properly

Note that if there is no overlap between requested ranges, Gradle will continue to choose the newest version that fits any range. The way to avoid this is to ask Gradle to fail on version conflicts, as mentioned. (But overlapping ranges are no longer considered version conflicts).

We will soon be introducing dependencies declared with `strict` versions: these will enforce a particular version or range for a particular dependency, failing if not met. This will give you the power to "fail on version conflict" only for those versions you really care about.

It would be great if you could try out `4.3.1` to see if the improved behaviour helps you out. However, if you have non-overlapping ranges you'll still hit the same problem, in which case you'll need to wait for `strict` dependencies in a release soon. #3311 

We are planning large-scale improvements to dependency resolution in the coming months. Since there's no specific action we can take related to this issue, I hope it's OK that I close it. If you're interested in exploring further the idea of allowing different resolved graphs for compile and testing, please open another issue. (Note that not everyone is as diligent with version ranges as your team appears to be: many people will have `[1.0,)` in their declared version range long after `1.0` is no longer a valid target for compile.)

**Comment by xtracoder** [comment]
Thanks, we'll check how it works in near future.
